src/Composer/DependencyResolver/Problem.php

```diff
@@ -262,6 +263,17 @@ public static function getMissingPackageReason(RepositorySet $repositorySet, Req
             }
         }
 
+        if ($constraint instanceof Constraint && $constraint->getOperator() === Constraint::STR_OP_EQ && Preg::isMatch('{^dev-.*#.*}', $constraint->getPrettyString())) {
+            $newConstraint = Preg::replace('{ +as +([^,\s|]+)$}', '', $constraint->getPrettyString());
+            $packages = $repositorySet->findPackages($packageName, new MultiConstraint([
+                new Constraint(Constraint::STR_OP_EQ, $newConstraint),
+                new Constraint(Constraint::STR_OP_EQ, str_replace('#', '+', $newConstraint))
+            ], false));
+            if (\count($packages) > 0) {
+                return ["- Root composer.json requires $packageName".self::constraintToText($constraint) . ', ', 'found '.self::getPackageList($packages, $isVerbose, $pool, $constraint).' but '.(self::hasMultipleNames($packages) ? 'these' : 'it').' had a # in the branch name which was replaced by +. Make sure to require it as "'.str_replace('#', '+', $constraint->getPrettyString()).'".'];
```

Comment:
```suggestion
                return ["- Root composer.json requires $packageName".self::constraintToText($constraint) . ', ', 'found '.self::getPackageList($packages, $isVerbose, $pool, $constraint).' but '.(self::hasMultipleNames($packages) ? 'these' : 'it').'. The # character in branch names is replaced by a + character. Make sure to require "'.str_replace('#', '+', $constraint->getPrettyString()).'".'];
```

It's not really "but" cause the found thing lists the branch with a +